File: program/leslie.cpp
// State-Space Leslie Model

#include <TMB.hpp>
#include <iostream>

template<class Type>
Type objective_function<Type>::operator() ()
{
  // DATA //
  DATA_VECTOR(CPUE);
  DATA_VECTOR(CAT);
  DATA_VECTOR(WEEK);  
  DATA_IVECTOR(YEAR);
  DATA_IVECTOR(START);
  DATA_IVECTOR(END);
  DATA_INTEGER(Y);
  DATA_SCALAR(M);
  DATA_SCALAR(x_lo);
  DATA_SCALAR(x_up);
  DATA_VECTOR(nodes);
  DATA_VECTOR(wt);
  DATA_SCALAR(g);
  DATA_INTEGER(model);     // 0: HS, 1: MR, 2: BHS, 3: BH, 4: RI
  
  // PARAMETER //
  PARAMETER(log_a);
  PARAMETER(log_b);
  PARAMETER(tilde_n0);
  PARAMETER(log_tilde_q);
  PARAMETER(log_sigma);
  PARAMETER(log_tau);
  PARAMETER(log_eta);
  PARAMETER_VECTOR(log_q);  
  PARAMETER_VECTOR(n0);
  
  // PARAMETER TRANSFORMATION //
  Type a = exp(log_a);
  Type sigma = exp(log_sigma);
  Type tau = exp(log_tau);
  Type eta = exp(log_eta);
  int T = CPUE.size();
  int ng = wt.size();
  
  vector<Type> q(Y);
  vector<Type> n_last(Y);
  vector<Type> N_S(Y);
  vector<Type> N_E(Y);
  vector<Type> pred_n0(Y);
  vector<Type> U(T);
  vector<Type> n(T);
  vector<Type> N(T);
  vector<Type> pred_n(T);
     
  Type nll = 0.0;
  
  // GL integration
  
  Type xm_x = Type(0.5)*(x_up+x_lo);
  Type xr_x = Type(0.5)*(x_up-x_lo);
  vector<Type> dx_gi_x = xr_x*nodes;
  vector<Type> xp_x = xm_x+dx_gi_x;
  
  // DeLury Model
  
  for (int i=0;i<T;i++){
    if (START(i)==1) {
      if (i==0) nll += -dnorm(log_q(YEAR(i)),log_tilde_q,eta,true); 
      if (i > 0) nll += -dnorm(log_q(YEAR(i)),log_q(YEAR(i)-1),eta,true);
      n(i) = n0(YEAR(i))-M*Type(7.0)*WEEK(i);
      N(i) = exp(n(i));
      N_S(YEAR(i)) = exp(n0(YEAR(i)));
      U(i) = CAT(i)/N(i);
    }
    if (START(i)==0) {
      n(i) = n(i-1)-M*Type(7.0)*(WEEK(i)-WEEK(i-1))+log(Type(1.0)-U(i-1));
      N(i) = exp(n(i));
      U(i) = CAT(i)/N(i);
    }
    U(i) = CppAD::CondExpLe(Type(1.0)-U(i),Type(0.0),Type(0.99),U(i));
    if (END(i)==1){
      n_last(YEAR(i)) = n(i)-M*Type(7.0)*(Type(26.0)-WEEK(i))+log(Type(1.0)-U(i));
      N_E(YEAR(i)) = exp(n_last(YEAR(i)));
    }

    pred_n(i) = log_q(YEAR(i))+n(i);
    nll += -dnorm(log(CPUE(i)),pred_n(i),tau,true);
  }
  
  // Density-Dependent Model
  
  vector<Type> S = exp(n_last);
  Type max_S = max(S);
  Type min_S = min(S);
  Type pred;
  
  Type b = CppAD::CondExpLe(exp(log_b), min_S, min_S, exp(log_b));
  b = CppAD::CondExpGe(exp(log_b), max_S, max_S, exp(log_b));
  
  nll += -dnorm(n0(0),tilde_n0,sigma,true);
  for (int i=1;i<Y;i++){ 
    if (model==0) pred_n0(i) = CppAD::CondExpGe(S(i-1), b, log_a + log(b), log_a+log(S(i-1)));
    if (model==1) pred_n0(i) = log_a+log(S(i-1)+sqrt(pow(b,Type(2.0))+pow(g,Type(2.0))/Type(4.0))-sqrt(pow(S(i-1)-b,Type(2.0))+pow(g,Type(2.0))/Type(4.0)));
    if (model==2) {
      pred = 0.0;
      for (int j=0;j<ng;j++){
        pred += xr_x*wt(j)*a*b*pow(S(i-1)/b,1-pow(S(i-1)/b,1/xp_x(j)));
      }
      pred /= 2.0*xr_x;
      pred_n0(i) = CppAD::CondExpGe(S(i-1), b, log_a + log(b), log(pred));
    }
    if (model==3) pred_n0(i) = log_a + log(S(i-1)) - log(Type(1.0)+S(i-1)/b);
    if (model==4) pred_n0(i) = log_a + log(S(i-1)) - S(i-1)/b;
    nll += -dnorm(n0(i), pred_n0(i), sigma, true);
  }
  
  q = exp(log_q);
  
  Type n0_new = 0.0;
  if (model==0) n0_new = CppAD::CondExpGe(S(Y-1), b, log_a + log(b), log_a+log(S(Y-1)));
  if (model==1) n0_new = log_a+log(S(Y-1)+sqrt(pow(b,Type(2.0))+pow(g,Type(2.0))/Type(4.0))-sqrt(pow(S(Y-1)-b,Type(2.0))+pow(g,Type(2.0))/Type(4.0)));
  if (model==2) {
    pred=0.0;
    for (int j=0;j<ng;j++){
        pred += xr_x*wt(j)*a*b*pow(S(Y-1)/b,1-pow(S(Y-1)/b,1/xp_x(j)));
    }
    pred /= 2.0*xr_x;
    n0_new = CppAD::CondExpGe(S(Y-1), b, log_a + log(b), log(pred));
  }
  if (model==3) n0_new = log_a + log(S(Y-1)) - log(Type(1.0)+S(Y-1)/b);
  if (model==4) n0_new = log_a + log(S(Y-1)) - S(Y-1)/b;
  
  // ADREPORTS
  
  REPORT(a);
  REPORT(b);
  REPORT(max_S);
  REPORT(min_S);
          
  ADREPORT(a);
  ADREPORT(b);
  ADREPORT(sigma);
  ADREPORT(tau);
  ADREPORT(eta);
  ADREPORT(q);
  ADREPORT(U);
  ADREPORT(N_S);
  ADREPORT(N_E);
  ADREPORT(pred_n);
  ADREPORT(N);
  ADREPORT(n0_new);
    
  return nll;
}

File: program/leslie.R
### State-Space Leslie Removal Method with the BHS SR function

## Population dynamics model

leslie <- function(
  dat,
  parms=NULL,
  model=2,
  do_compile=TRUE,
  tmb_file="leslie_r3",
  M=0.0133,     # Roa-Ureta & Arkhipkin (2007)
  g0=0.01,
  x_int=c(0,5),
  n_g=10,
  a_range=c(1,10),
  b_range=c(1,10),
  a_init=NULL,
  b_init=NULL,
  max_P=0.5,
  maxit=1000,
  maxeval=1000,
  silent=TRUE,
  pre_process=TRUE,
  phase_a=FALSE,
  phase_b=FALSE,
  no_est=FALSE,
  make_plot=TRUE,
  optimizer="nlminb",
  method="BFGS"
){
  require(TMB)
  require(tidyverse)
  require(statmod)

  # Gauss-Legendre integration
  
  g_le <- gauss.quad(n_g,kind="legendre")
  
  nodes <- g_le$nodes
  wt <- g_le$weights
  
  #

  if (pre_process){
    dat <- dat %>% mutate(Cat=Catch/Mass, CR=CPUE/Mass)
    TC <- dat %>% group_by(Year) %>% summarize(TC=sum(Cat))
    Years <- unique(dat$Year)
    leslie_res <- lapply(Years, function(i) lm(CR~cumsum(Cat),data=subset(dat, Year==i)))
    lm_res <- sapply(1:length(leslie_res), function(i) c(-leslie_res[[i]]$coef[2], -leslie_res[[i]]$coef[1]/leslie_res[[i]]$coef[2]))
    lm_res <- rbind(lm_res, TC$TC)
    lm_res <- as.data.frame(t(lm_res))
    names(lm_res) <- c("q","N0","Catch")
    rownames(lm_res) <- Years
  }
  
  if (do_compile){
    compile(paste0(tmb_file,".cpp"))
    dyn.load(dynlib(tmb_file))
  }
  
  model_name <- c("HS","MR","BHS","BH","RI")
  
  dat1 <- dat %>% mutate(Y=Year-min(Year)) %>% group_by(Year) %>% mutate(W=Week-min(Week), START=(W==0), END=(W==max(W)))
  dat1c <- dat %>% group_by(Year) %>% summarize(CC=sum(Cat))
  nY <- max(dat1$Y)+1
  dat_bhs <- list(CPUE=dat1$CR, CAT=dat1$Cat, WEEK=dat1$Week, YEAR=dat1$Y, START=as.numeric(dat1$START), END=as.numeric(dat1$END), Y=nY, M=M, x_lo=x_int[1], x_up=x_int[2], nodes=nodes, wt=wt, g=g0, model=model)
  if (is.null(parms)){
    par_bhs <- list(
      log_a=2,
      log_b=2,
      tilde_n0=max(mean(log(lm_res$N0)), max(log(dat1c$CC*1.5))),
      log_tilde_q=mean(log(lm_res$q)),
      log_sigma=log(0.3),
      log_tau=log(0.3),
      log_eta=log(0.3),
      log_q=log(lm_res$q),
      n0=pmax(log(lm_res$N0), log(dat1c$CC*1.5))
    )
  } else par_bhs <- parms
  
  if (!is.null(a_init)) par_bhs$log_a <- a_init
  if (!is.null(b_init)) par_bhs$log_b <- b_init

  if (par_bhs$log_b < min(par_bhs$n0)) par_bhs$log_b <- 1.1*min(par_bhs$n0)
  if (par_bhs$log_b > max_P*max(par_bhs$n0)) par_bhs$log_b <- max_P*(max(par_bhs$n0)+0.1*par_bhs$log_b)
 
  lo <- c(a_range[1],b_range[1],rep(-20,5))
  up <- c(a_range[2],b_range[2],rep(20,5))
  
  maps <- list()
  if (phase_a) maps$log_a <- factor(NA)
  if (phase_b) maps$log_b <- factor(NA)
  obj <- MakeADFun(dat_bhs, par_bhs, map=maps, random=c("log_q","n0"), DLL=tmb_file, silent=silent)
  kk <- 1
  dd <- c(seq(0.01,0.5,len=5),seq(-0.5,-0.01,len=5))
  while(is.na(as.numeric(obj$fn())) & kk <= 10){
    par_bhs$log_b <- par_bhs$log_b+dd[kk]
    obj <- MakeADFun(dat_bhs, par_bhs, map=maps, random=c("log_q","n0"), DLL=tmb_file, silent=silent)
    kk <- kk + 1
  }
  if (no_est) mod_bhs <- list(par=obj$par, convergence=1) else {
    if (optimizer=="optim") {
        if (method=="L-BFGS-B") mod_bhs <- optim(obj$par, obj$fn, obj$gr, method=method, control=list(maxit=maxit), lower=lo, upper=up) else mod_bhs <- optim(obj$par, obj$fn, obj$gr, method=method, control=list(maxit=maxit))
    }
    if (optimizer=="nlminb") mod_bhs <- nlminb(obj$par, obj$fn, obj$gr, lower=lo, upper=up, control=list(iter.max=maxit, eval.max=maxeval))
    if (phase_a){
      par_bhs <- obj$env$parList(mod_bhs$par)
      maps$log_a <- NULL
      obj <- MakeADFun(dat_bhs, par_bhs, map=maps, random=c("log_q","n0"), DLL=tmb_file, silent=silent) 
      if (optimizer=="optim") {
        if (method=="L-BFGS-B") mod_bhs <- optim(obj$par, obj$fn, obj$gr, method=method, control=list(maxit=maxit), lower=lo, upper=up) else mod_bhs <- optim(obj$par, obj$fn, obj$gr, method=method, control=list(maxit=maxit))
      }
      if (optimizer=="nlminb") mod_bhs <- nlminb(obj$par, obj$fn, obj$gr, lower=lo, upper=up, control=list(iter.max=maxit, eval.max=maxeval))
    }
    if (phase_b){
      par_bhs <- obj$env$parList(mod_bhs$par)
      maps$log_b <- NULL
      obj <- MakeADFun(dat_bhs, par_bhs, map=maps, random=c("log_q","n0"), DLL=tmb_file, silent=silent) 
      if (optimizer=="optim") {
        if (method=="L-BFGS-B") mod_bhs <- optim(obj$par, obj$fn, obj$gr, method=method, control=list(maxit=maxit), lower=lo, upper=up) else mod_bhs <- optim(obj$par, obj$fn, obj$gr, method=method, control=list(maxit=maxit))
      }
      if (optimizer=="nlminb") mod_bhs <- nlminb(obj$par, obj$fn, obj$gr, lower=lo, upper=up, control=list(iter.max=maxit, eval.max=maxeval))
    }
  }
  
  sdrep <- sdreport(obj)
  
  conv_diag <- c("mod_conv" = (mod_bhs$convergence==0), "grad" = (max(abs(sdrep$gradient.fixed)) < 10^(-2)), "pdHess" = (sdrep$pdHess==TRUE), "vcov" = (all(is.na(sdrep$cov.fixed))==FALSE))
  conv_diag <- c(conv_diag, "convergence"=all(conv_diag))
  
  if (rev(mod_bhs$par)[1] < -10){
    par_bhs$log_eta <- -10
    maps$log_eta <- factor(NA)
    lo <- lo[1:(length(lo)-1)]
    up <- up[1:(length(up)-1)]    
    obj <- MakeADFun(dat_bhs, par_bhs, map=maps, random=c("log_q","n0"), DLL=tmb_file, silent=silent)
    if (optimizer=="optim") {
      if (method=="L-BFGS-B") mod_bhs <- optim(obj$par, obj$fn, obj$gr, method=method, control=list(maxit=maxit), lower=lo, upper=up) else mod_bhs <- optim(obj$par, obj$fn, obj$gr, method=method, control=list(maxit=maxit))
    }
    if (optimizer=="nlminb") mod_bhs <- nlminb(obj$par, obj$fn, obj$gr, lower=lo, upper=up, control=list(iter.max=maxit, eval.max=maxeval))
    
    sdrep <- sdreport(obj)
    conv_diag <- c("mod_conv" = (mod_bhs$convergence==0), "grad" = (max(abs(sdrep$gradient.fixed)) < 10^(-2)), "pdHess" = (sdrep$pdHess==TRUE), "vcov" = (all(is.na(sdrep$cov.fixed))==FALSE))
    conv_diag <- c(conv_diag, "convergence"=all(conv_diag))
  }

  pars <- obj$env$parList(mod_bhs$par)
  pars$a <- obj$report()$a
  pars$b <- obj$report()$b
  
  xm_x <- 0.5*(x_int[2]+x_int[1])
  xr_x <- 0.5*(x_int[2]-x_int[1])
  dx_gi_x <- xr_x*nodes
  xp_x <- xm_x+dx_gi_x
  
  GL <- list(xr_x=xr_x, xp_x=xp_x, wt=wt)

  N_E <- summary(sdrep)[rownames(summary(sdrep))=="N_E",1]
  N_S <- summary(sdrep)[rownames(summary(sdrep))=="N_S",1]
  
  ts_out <- data.frame(S=N_E, R=N_S)
  
  if (make_plot){
    X <- seq(0,round(max(N_E)*1.1))
    pred_out <- data.frame(ES = X, ER = pred_R(X, pars, GL, g0, model=model))

    p1 <- ggplot(ts_out, aes(x=S,y=R))+geom_point()+labs(x="S", y="R")+theme_bw()+geom_line(data=pred_out, aes(x=ES,y=ER),color="blue")
  } else p1 <- NULL
  
  list(model=model, model_name=model_name[model+1], M=M, x_int=x_int, n_g=n_g, GL=GL, no_est=no_est, g=g0, phase_b=phase_b, conv_diag=conv_diag, p=p1, obj=obj, mod=mod_bhs, pars=pars, dat=dat_bhs, sdrep=sdrep, ts_out=ts_out)
}

## Prediction

pred_R <- function(S,p,GL,g=0.01,model=0) {
   a <- p$a
   b <- p$b
   
   if (model==0) res <- ifelse(S < b, a*S, a*b) 
   if (model==1) res <- a*(S+sqrt(b^2+g^2/4)-sqrt((S-b)^2+g^2/4))
   if (model==2){
     pred <- function(S) sum(GL$xr_x*GL$wt*a*b*(S/b)^(1-(S/b)^(1/GL$xp_x)))/(2*GL$xr_x)
     pred <- Vectorize(pred, "S")
     res <- ifelse(S < b, pred(S), a*b)
   }
   if (model==3) res <- a*S/(1+S/b)
   if (model==4) res <- a*S*exp(-S/b)
   
   return(res)
}

SR_plot <- function(res, msy_res=NULL){
 ts_out <- res$ts_out
 model <- res$model
 pars <- res$par
 GL <- res$GL
 g <- res$g
 
 X <- seq(0,round(max(ts_out$S)*1.1))
 pred_out <- data.frame(ES = X, ER = pred_R(X, pars, GL, g=g, model=model))

 p1 <- ggplot(ts_out, aes(x=S,y=R))+geom_point()+labs(x="S", y="R")+theme_bw()+geom_line(data=pred_out, aes(x=ES,y=ER),color="blue")
 
 p1
}

## MSY estimation

n_est <- function(n, p, M, GL, g, U=0.5, n_g=50, stochastic=TRUE, model=2){
   require(statmod)
   g_he <- gauss.quad(n_g,kind="hermite")

   if (stochastic) sigma <- exp(p$log_sigma) else sigma <- rep(0, n_g)
   
   S <- function(z) exp(n-M*7*26+log(1-U)+z)
   integrand <- function(z) pred_R(S(z), p, GL, g, model=model)
   n_pred <- sum(g_he$weights*integrand(sqrt(2)*sigma*g_he$nodes))/sqrt(pi)
   
   return(log(n_pred))
}

n_finder <- function(x, res, U=0.5, st=TRUE){
  model <- res$model
  p <- res$par
  GL <- res$GL
  g <- res$g
  M <- res$M
 
 (x - n_est(x, p, M, GL, g, U=U, stochastic=st, model=model))^2
}

sy <- function(res, U=0.5, st=TRUE, n_range=c(2,12)) {
  n_eq <- optimize(n_finder, n_range, res=res, U=U, st=st)
  res <- log(U)+n_eq$minimum
  attr(res, "obj") <- n_eq$objective
  res
}

msy_est <- function(
  res,
  ST=TRUE,
  n_range=c(2,12),
  U_range=c(0.01,0.99)
){
  model <- res$model
  M <- res$M
  p <- res$par
  
  U_msy <- optimize(sy, U_range, maximum=TRUE, res=res, st=ST, n_range=n_range)
  n_msy <- optimize(n_finder, n_range, res=res, U=U_msy$maximum, st=ST)
  S_msy <- exp(n_msy$minimum)*exp(-M*7*26)*(1-U_msy$maximum)

  list(model=model, model_name=res$model_name, p=p, U_msy=U_msy, n_msy=n_msy, S_msy=S_msy, MSY=U_msy$maximum*exp(n_msy$minimum))
}

##

sim_est <- function(res, U=0.5, d=1, Sim=3000, Y=100, seed0=1){
  set.seed(seed0)
  model <- res$model
  p <- res$par
  M <- res$M
  GL <- res$GL
  g <- res$g
  n0 <- p$log_a+p$log_b
  sigma <- exp(p$log_sigma)
  
  n_vec <- matrix(NA, nrow=Y, ncol=Sim)

  z <- matrix(rnorm(Y*Sim), nrow=Y, ncol=Sim)

  n_vec[1,] <- n0+sigma*z[1,]
  
  for (i in 1:(Y-1)){
    n_vec[i+1,] <- log(pred_R(exp(n_vec[i,]-M*7*26+log(1-U)),p,GL,g,model))+sigma*z[i+1,]
  }
  
  exp(n_vec+log(U)-d*sigma^2/2)
}

##

msy_sim_est <- function(
  Sim_dat,
  res,
  ST=TRUE,
  n_range=c(5,10),
  U_range=c(0.01,0.99)
){
  model <- Sim_dat$model
  M <- res$M
  GL <- res$GL
  g <- res$g
  
  m <- Sim_dat$m
  
  res1 <- list(model=model, M=M, GL=GL, g=g)
  
  Res <- list()
  for (i in 1:m){
    p <- as.list(Sim_dat$parms[i,])
    p$a <- exp(p$log_a)
    p$b <- exp(p$log_b)
    res1$par <- p
  
    U_msy <- optimize(sy, U_range, maximum=TRUE, res=res1, st=ST, n_range=n_range)
    n_msy <- optimize(n_finder, n_range, res=res1, U=U_msy$maximum, st=ST)
    S_msy <- exp(n_msy$minimum)*exp(-M*7*26)*(1-U_msy$maximum)

    Res[[i]] <- list(model=model, model_name=res$model_name, p=p, U_msy=U_msy, n_msy=n_msy, S_msy=S_msy, MSY=U_msy$maximum*exp(n_msy$minimum))
  }
  
  Res
}

## Convergence Comparison

comp_table <- function(Res0, Res1){
  table(sapply(1:100, function(i) Res0[[i]]$conv_diag[4]), sapply(1:100, function(i) Res1[[i]]$conv_diag[4]))
}
